lib/twilio/audioplayer/audioplayer.ts

```diff
@@ -0,0 +1,363 @@
+/**
+ * @packageDocumentation
+ * @module Voice
+ * @internalapi
+ */
+// @ts-nocheck
+import Deferred from './deferred';
+import EventTarget from './eventtarget';
+
+import ChromeAudioContext, { ChromeHTMLAudioElement, MediaStreamAudioDestinationNode } from './chromeaudiocontext';
+
+/**
+ * Options that may be passed to AudioPlayer for dependency injection.
+ */
+interface IAudioPlayerOptions {
+  /**
+   * The factory for Audio.
+   */
+  AudioFactory: any;
+
+  /**
+   * The factory for XMLHttpRequest.
+   */
+  XMLHttpRequestFactory: any;
+}
+
+/**
+ * An {@link AudioPlayer} is an HTMLAudioElement-like object that uses AudioContext
+ *   to circumvent browser limitations.
+ * @private
+ */
+class AudioPlayer extends EventTarget {
+  /**
+   * The AudioContext. This is passed in at construction and used to create
+   *   MediaStreamBuffers and AudioNodes for playing sound through.
+   */
+  private _audioContext: ChromeAudioContext;
+
+  /**
+   * The Audio element that is used to play sound through when a non-default
+   *   sinkId is set.
+   */
+  private _audioElement: ChromeHTMLAudioElement;
+
+  /**
+   * The AudioBufferSourceNode of the actively loaded sound. Null if a sound
+   *   has not been loaded yet. This is re-used for each time the sound is
+   *   played.
+   */
+  private _audioNode: AudioBufferSourceNode|null = null;
+
+  /**
+   * A Promise for the AudioBuffer. Listening for the resolution of this Promise
+   *   delays an operation until after the sound is loaded and ready to be
+   *   played.
+   */
+  private _bufferPromise: Promise<AudioBuffer>;
+
+  /**
+   * The current destination for audio playback. This is set to context.destination
+   *   when default, or a specific MediaStreamAudioDestinationNode when setSinkId
+   *   is set.
+   */
+  private _destination: MediaStreamAudioDestinationNode;
+
+  /**
+   * The GainNode used to control whether the sound is muted.
+   */
+  private _gainNode: GainNode;
+
+  /**
+   * Whether or not the audio element should loop. If disabled during playback,
+   *   playing continues until the sound ends and then stops looping.
+   */
+  private _loop: boolean = false;
+
+  /**
+   * An Array of deferred-like objects for each pending `play` Promise. When
+   *   .pause() is called or .src is set, all pending play Promises are
+   *   immediately rejected.
+   */
+  private _pendingPlayDeferreds: Array<Deferred<AudioBuffer>> = [];
+
+  /**
+   * The current sinkId of the device audio is being played through.
+   */
+  private _sinkId: string = 'default';
+
+  /**
+   * The source URL of the sound to play. When set, the currently playing sound will stop.
+   */
+  private _src: string = '';
+
+  /**
+   * The Factory to use to construct an XMLHttpRequest.
+   */
+  private _XMLHttpRequest: any;
+
+  get destination(): MediaStreamAudioDestinationNode { return this._destination; }
```

Comment:
Out of scope, but I prefer not to use `getter` and `setter` methods. It obfuscates the possibility of side-effects when accessing members.